### PR TITLE
Issue #7983: Resolve Pitest Issues - RegexpCheck (2)

### LINKED ISSUE #7983: Resolve Pitest Issues - RegexpCheck (2)
Child issue of https://github.com/checkstyle/checkstyle/issues/7797 ,

https://github.com/checkstyle/checkstyle/blob/f9131f9536c71d963c83aab1550b7a0371aa9ff5/.ci/pitest.sh#L53

### PR DESCRIPTION
**Pull request by anhminhtran235**
Fixes #7983 

<h1>1. Pitest Report</h1>

- Link: https://user-images.githubusercontent.com/54651212/95527351-e14cac00-0989-11eb-9f3d-f851614faa7f.PNG
- Surviving Mutations:
removed conditional - replaced equality check with false → SURVIVED
removed conditional - replaced equality check with true → SURVIVED

<h1>2. Mutate Branch</h1>

- Hardcoded Mutated: https://user-images.githubusercontent.com/54651212/95529357-8cac2f80-098f-11eb-8a57-01830d3acfc9.PNG

<h1>3. Regression Report</h1>

Not needed due to no main code change

<h1>4. Code logic and analysis</h1>

Similar to #8001, `(ignore || illegalPattern || checkForDuplicates)` should be removed from

```
  private boolean canContinueValidation(boolean ignore) {
        return errorCount <= errorLimit - 1
                && (ignore || illegalPattern || checkForDuplicates);
    }
```

The only way for `(ignore || illegalPattern || checkForDuplicates)` to be false is if all of them are false. This means the pattern is `not illegal`, and the user is `not checking for duplicates`. If this is the case, the program **only needs to find 1 valid match** and then it can  _stop right away_. The `canContinueValidation()` function above **correctly check this**. However, mutations are still alive because if we substitute `(ignore || illegalPattern || checkForDuplicates)` with `true`, the program _does **not** stop_ but **nothing went wrong** (Although the program runs slower since it has to keep going through the rest of the file).

If we only substitute `(ignore || illegalPattern || checkForDuplicates)` with `true` (i.e remove it completely), we kill all mutations BUT now the program will _**not** stop early_ and instead go through the whole file even if it only needs to find 1 match. I think that is what causes `timeout` in #8001: Check whole file => Recursion stack is too deep.

I tried to remove `(ignore || illegalPattern || checkForDuplicates)` and added logic like this to stop early:

```
    final boolean onlyNeedOneMatch = !illegalPattern && !checkForDuplicates;
    if (foundOneMatch && onlyNeedOneMatch) return;
```

That doesn't work either, mutations are still alive since mutating `foundOneMatch && onlyNeedOneMatch` to `false` doesn't break the program (only make it slower).

=> So in this PR, I **turn the recursive call to a while loop**, hoping to get rid of `timeout` issue.

Pitest report after fix: kill all mutations
https://user-images.githubusercontent.com/54651212/95529213-33dc9700-098f-11eb-9edd-3acfaab254c8.PNG

### CODE REVIEW

**Comment by anhminhtran235** [comment]
@rnveach @romani Lots of checks are failing. That doesn't look good. Do you guys know why?

**Comment by romani** [comment]
Please look at all of them, all of them correct problems.

**Comment by romani** [comment]
```
Command: ./.ci/wercker.sh no-error-checkstyles-sevntu
....
[INFO] --- maven-checkstyle-plugin:3.1.1:check (sevntu-checkstyle-check) @ checkstyle ---
[INFO] There is 1 error reported by Checkstyle 8.37-SNAPSHOT with /pipeline/source/config/checkstyle_sevntu_checks.xml ruleset.
[ERROR] src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java:[629,31] (extension) LogicConditionNeedOptimization: Condition with && at line 629 position 30 need optimization. All method calls are advised to move to end of logic expression.

```

**Comment by anhminhtran235** [comment]
> ```
> Command: ./.ci/wercker.sh no-error-checkstyles-sevntu
> ....
> [INFO] --- maven-checkstyle-plugin:3.1.1:check (sevntu-checkstyle-check) @ checkstyle ---
> [INFO] There is 1 error reported by Checkstyle 8.37-SNAPSHOT with /pipeline/source/config/checkstyle_sevntu_checks.xml ruleset.
> [ERROR] src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java:[629,31] (extension) LogicConditionNeedOptimization: Condition with && at line 629 position 30 need optimization. All method calls are advised to move to end of logic expression.
> ```

Fixed. I saw you ran command `./.ci/wercker.sh no-error-checkstyles-sevntu`. Are there any other commands I should run to check if my code is correct?

**Comment by romani** [comment]
@anhminhtran235 , do not hesitate to click on  "Details" link of each failed CI and read logs of failure, it will let you know what is wrong with your code. example: https://travis-ci.org/github/checkstyle/checkstyle/builds/734184491

only wercker is not accessble to review, so I shared output.

**Comment by rnveach** [comment]
My notes on the original attempt to resolve this pitest is at https://github.com/checkstyle/checkstyle/pull/8001#issuecomment-631984819 .

**Comment by anhminhtran235** [comment]
> @anhminhtran235 , do not hesitate to click on "Details" link of each failed CI and read logs of failure, it will let you know what is wrong with your code. example: https://travis-ci.org/github/checkstyle/checkstyle/builds/734184491
> 
> only wercker is not accessble to review, so I shared output.

@romani Thank you. I have looked at the "Details" link and fixed those issues. Wercker still fails though. Could you take a look? I suspect changing recursion to while loop doesn't help?

**Comment by anhminhtran235** [comment]
> My notes on the original attempt to resolve this pitest is at [#8001 (comment)](https://github.com/checkstyle/checkstyle/pull/8001#issuecomment-629746934) and [#8001 (comment)](https://github.com/checkstyle/checkstyle/pull/8001#issuecomment-631984819) .
> 
> Beware of the timeout issue that we ran into in PR. It is not a guarantee that it will always timeout.

@rnveach @romani I have looked at rnveach's comment and attempt to fix. This is my thoughts. Please correct me if I'm wrong:
- I think if we still want to stop early, we need to **add more test cases** to kill mutation. This is because _without_ stopping early, the program still outputs correctly => no test case fails => When Pitest mutates our code from `if (readyToStop) stopEarly();` to `if (false) stopEarly();` nothing goes wrong => **Mutation survives as long as we stop early**
- Removing `(ignore || illegalPattern || checkForDuplicates)` causes the program **not** stopping early => kill all mutations **but** timeout.

I think we need 1 more test case: Either having a very, very long file to check, which will timeout / stackoverflow if not stop early; or add a specific message when stop early: `found the pattern, stop searching` and add test case for that.

**Comment by rnveach** [comment]
My only other thought would be to check if you can do anything by converting it to a while loop. That is what I was hoping to look in at some time. There is some stuff built into pitest to be more friendlier to certain kinds of loops.

**Comment by anhminhtran235** [comment]
@rnveach I have converted the recursion to a while loop. Wercker still fails though. Do you have any suggestions?

**Comment by romani** [comment]
Wercker failing at `Command: ./.ci/wercker.sh no-exception-hbase` please run this on your local to reproduce.
I restarted wercker to recheck stability

**Comment by rnveach** [comment]
I see didn't see hbase failure. If it is failing in exact same way as previous PR, which was a bad change, then your conversion is somehow flawed. I would expect only a pitest failure at the most from the conversion complaining about something in the new changes.

**Comment by romani** [comment]
right now wercker is failing to even start execution at all, not related to PR code. restarted.

**Comment by romani** [comment]
@anhminhtran235 , 
all pitest reports are available at circle-ci - https://app.circleci.com/pipelines/github/checkstyle/checkstyle/3400/workflows/87a0a854-5efd-40ed-b2db-6b4b5aa4d194/jobs/64366/artifacts
we and you can always review them

> 3. Regression Report

please do it like https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#report-generation it is easy and you should use all projects(that we have in config) to execute on. Please do.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java [hunk 1]
```diff
@@ -625,8 +625,8 @@ public void beginTree(DetailAST rootAST) {
 
     /** Recursive method that finds the matches. */
     private void findMatch() {
-        final boolean foundMatch = matcher.find();
-        if (foundMatch) {
+        boolean cont = true;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java:628]: cont -> shouldContinue

**Review by romani — CHANGES_REQUESTED** [review]
please update your description to match your latest code. 

minor:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java [hunk 2]
```diff
@@ -641,24 +641,23 @@ private void findMatch() {
                     logMessage(startLine);
                 }
             }
-            if (canContinueValidation(ignore)) {
-                findMatch();
+            if (!canContinueValidation()) {
+                cont = false;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java:645]: can it be `shouldContinue = canContinueValidation()` ?

**Review by romani — CHANGES_REQUESTED** [review]
item:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java (hunk 1 above)
Comment (anhminhtran235) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java]: fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java (hunk 2 above)
Comment (anhminhtran235) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java]: fixed. I removed shouldContinue completely. It still passes all test cases

**Comment by anhminhtran235** [comment]
Diff Regression config: https://gist.githubusercontent.com/anhminhtran235/b466939c3780cc9bc30c683ebec8a318/raw/bfed3bf7f78ca38df4b2c4cdaf616adc3fd97d3f/confix.xml
Diff Regression projects: https://gist.githubusercontent.com/anhminhtran235/5285bf25123afb11527305d10ffc6b2b/raw/9dc09fc1df29ff8655bc9c7fdd278063e2d5236d/projects-to-test-on.properties

**Comment by anhminhtran235** [comment]
Github, generate report

**Comment by anhminhtran235** [comment]
Diff Regression config: https://gist.githubusercontent.com/anhminhtran235/7fb296bc2f561420f05d0351bc74f270/raw/100ba8f52a6eabbf847843d0c10cdd3b2f15bfa4/config.xml
Diff Regression projects: https://gist.githubusercontent.com/anhminhtran235/71bf6a64f93e1d8745084e173b6ea6d8/raw/4d0d16a67416bd141eeb240c78f06711eeccb03c/projects-to-test-on.properties

**Comment by anhminhtran235** [comment]
GitHub, generate report

**Comment by anhminhtran235** [comment]
@romani Ran `./.ci/wercker.sh no-exception-hbase` on my computer. Build success, no error.

**Comment by anhminhtran235** [comment]
> Github, generate report

@romani I tried to generate report, but I think I did something wrong. I don't understand how github can generate report given only 2 files. Do I need to connect my gist.github to my PR somehow?

**Comment by romani** [comment]
@anhminhtran235 , please put https://github.com/checkstyle/checkstyle/pull/8883#issuecomment-706770850 to PR description and it will work.

**Comment by romani** [comment]
Github, generate report

**Comment by anhminhtran235** [comment]
@romani @rnveach Keeping code the same, **add 1 more test case**, kill all mutations:
```
      // New test case:
      @Test
      public void testStopEarly() throws Exception {
          final DefaultConfiguration checkConfig =
              createModuleConfig(RegexpCheck.class);
          checkConfig.addAttribute("format", "a");
          checkConfig.addAttribute("illegalPattern", "false");
          checkConfig.addAttribute("duplicateLimit", "-1");
          final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
          verify(checkConfig, getPath("InputStopEarly.java"), expected);
      }
```
InputStopEarly.java: https://gist.github.com/anhminhtran235/ed51939a941b5dbb954d787047a7e34b

- I think this is what we should do, because removing `(ignore || illegalPattern || checkForDuplicates)` clearly slows the program down
- New test case kill all mutations because when pitest set `(ignore || illegalPattern || checkForDuplicates)` to `true`, the program doesn't stop => Stackoverflow

**Comment by romani** [comment]
>I think this is what we should do

Please do.

**Comment by anhminhtran235** [comment]
@romani 
Code fixed, test case added.

Could you please check what's wrong with Github report? I read the error but don't understand it. My `Diff Regression config` and `Diff Regression projects` are simply copied from these two demo links:

https://raw.githubusercontent.com/checkstyle/contribution/master/checkstyle-tester/projects-to-test-on.properties
https://gist.githubusercontent.com/strkkk/121653f4a334be38b9e77e4245e144e2/raw/691fe6e90ff40473707ce77518b7a0b058bd0955/config.xml

**Review by romani — APPROVED** [review]

**Comment by anhminhtran235** [comment]
@romani @rnveach My PR's having two errors:
```
All check input files in the 'src/(it|test)/resources*/' folder must have a comment at the top of the file
specifying the check configuration used to produce the violations, that begins with "/* Config:"
[ConfigCommentOnTopInputs]
```
and
```
C:\projects\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\regexp\regexp\InputStopEarly.java:0:
All check input files in the 'src/(it|test)/resources*/' folder must have trailing comments `// ok` and/or `// violation`.
[StyleValidationCommentInInputs]

```
[This test file](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/regexp/regexp/InputRegexpStartingWithEmptyLine.java) doesn't have `config` and `// ok` either, but has no error. I don't think my test file needs `// ok` or `// violation`. Is there a way to disable this error?

**Comment by romani** [comment]
>I don't think my test file needs // ok or // violation. Is there a way to disable this error?

Please put "ok" somewhere in input file as work around.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheckTest.java [hunk 3]
```diff
@@ -112,6 +112,17 @@ public void testIllegalPass() throws Exception {
...
 
+    @Test
+    public void testStopEarly() throws Exception {
+        final DefaultConfiguration checkConfig =
+            createModuleConfig(RegexpCheck.class);
+        checkConfig.addAttribute("format", "a");
+        checkConfig.addAttribute("illegalPattern", "false");
+        checkConfig.addAttribute("duplicateLimit", "-1");
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verify(checkConfig, getPath("InputStopEarly.java"), expected);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheckTest.java:123]: Name it InputRegexpCheckStopEarly

**Comment by strkkk** [comment]
> Could you please check what's wrong with Github report? I read the error but don't understand it. My Diff Regression config and Diff Regression projects are simply copied from these two demo links:

@anhminhtran235 you need to create your own config and projects files (for your changes in PR) and add links to them. You can do it with github gist, way to do it is described in same doc.

**Review by strkkk — CHANGES_REQUESTED** [review]
please fix CI issues

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheckTest.java (hunk 3 above)
Comment (anhminhtran235) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheckTest.java]: fixed

**Comment by anhminhtran235** [comment]
@rnveach I no longer edited the file, only added1 more test case. Do I need Regression Report? Or only Pitest Report is fine?

**Comment by romani** [comment]
>Do I need Regression Report?

No need for regression report if no main code changes.

**Comment by romani** [comment]
CI failure:
`[ERROR] [checkstyle] [ERROR] C:\projects\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\regexp\regexp\InputRegexpCheckStopEarly.java:0: All check input files in the 'src/(it|test)/resources*/' folder must                    have trailing comments "// ok" and/or "// violation". [StyleValidationCommentInInputs]`

**Comment by romani** [comment]
@strkkk , please review, this time CI should pass

**Review by strkkk — APPROVED** [review]

**Pull request merged**